PhysicsTools/NanoAOD/plugins/LeptonJetVarProducer.cc

```diff
@@ -0,0 +1,211 @@
+
+  for (uint il = 0; il<nLep; il++){
+    for (uint ij = 0; ij<nJet; ij++){
+      auto lep = srcLep->ptrAt(il);
+      auto jet = srcJet->ptrAt(ij);
+      if(matchByCommonSourceCandidatePtr(*lep,*jet)){
+	  auto res = calculatePtRatioRel(lep,jet,pv);
+	  ptRatio[il] = std::get<0>(res);
+	  ptRel[il] = std::get<1>(res);
+	  jetNDauChargedMVASel[il] = std::get<2>(res);
+	  jetForLepJetVar[il] = jet;
+	  break; // take leading jet with shared source candidates
+	}
+    }
+  }
+
+  std::unique_ptr<edm::ValueMap<float>> ptRatioV(new edm::ValueMap<float>());
+  edm::ValueMap<float>::Filler fillerRatio(*ptRatioV);
+  fillerRatio.insert(srcLep,ptRatio.begin(),ptRatio.end());
+  fillerRatio.fill();
+  iEvent.put(std::move(ptRatioV),"ptRatio");
+
+  std::unique_ptr<edm::ValueMap<float>> ptRelV(new edm::ValueMap<float>());
+  edm::ValueMap<float>::Filler fillerRel(*ptRelV);
+  fillerRel.insert(srcLep,ptRel.begin(),ptRel.end());
+  fillerRel.fill();
+  iEvent.put(std::move(ptRelV),"ptRel");
+
+  std::unique_ptr<edm::ValueMap<float>> jetNDauChargedMVASelV(new edm::ValueMap<float>());
+  edm::ValueMap<float>::Filler fillerNDau(*jetNDauChargedMVASelV);
+  fillerNDau.insert(srcLep,jetNDauChargedMVASel.begin(),jetNDauChargedMVASel.end());
+  fillerNDau.fill();
+  iEvent.put(std::move(jetNDauChargedMVASelV),"jetNDauChargedMVASel");
+
+  std::unique_ptr<edm::ValueMap<reco::CandidatePtr>> jetForLepJetVarV(new edm::ValueMap<reco::CandidatePtr>());
+  edm::ValueMap<reco::CandidatePtr>::Filler fillerjetForLepJetVar(*jetForLepJetVarV);
+  fillerjetForLepJetVar.insert(srcLep,jetForLepJetVar.begin(),jetForLepJetVar.end());
+  fillerjetForLepJetVar.fill();
+  iEvent.put(std::move(jetForLepJetVarV),"jetForLepJetVar");
+
+
+}
+
+template <typename T>
+std::tuple<float,float,float>
+LeptonJetVarProducer<T>::calculatePtRatioRel(edm::Ptr<reco::Candidate> lep, edm::Ptr<pat::Jet> jet, const reco::Vertex &vtx) const {
+ 
+  auto rawp4_ = jet->correctedP4("Uncorrected");
+  auto rawp4 = TLorentzVector(rawp4_.pt(),rawp4_.eta(),rawp4_.phi(),rawp4_.energy());
+  auto lepp4 = TLorentzVector(lep->pt(),lep->eta(),lep->phi(),lep->energy());
+
+  if ((rawp4-lepp4).Rho()<1e-4) return std::tuple<float,float,float>(1.0,0.0,0.0);
+
+  auto jetp4 = (rawp4 - lepp4*(1.0/jet->jecFactor("L1FastJet")))*(jet->pt()/rawp4.Pt())+lepp4;
+  auto ptratio = lepp4.Pt()/jetp4.Pt();
+  auto ptrel = lepp4.Perp((jetp4-lepp4).Vect());
+
+  unsigned jndau = 0;
+  for(const auto _d : jet->daughterPtrVector()) {
+    const auto d = dynamic_cast<const pat::PackedCandidate*>(_d.get());
+    if (d->charge()==0) continue;
+    if (d->fromPV()<=1) continue;
+    if (deltaR(*d,*lep)>0.4) continue;
+    if (!(d->hasTrackDetails())) continue;
+    auto tk = d->pseudoTrack();
+    if(tk.pt()>1 &&
+       tk.hitPattern().numberOfValidHits()>=8 &&
+       tk.hitPattern().numberOfValidPixelHits()>=2 &&
+       tk.normalizedChi2()<5 &&
+       fabs(tk.dxy(vtx.position()))<0.2 &&
+       fabs(tk.dz(vtx.position()))<17
```

Comment:
@arizzi 
given that fromPV is required to be >=2, does this dz cut make sense?
Also, in PhotonIDValueMapProducer.cc you have switched to calls to PackedCandidates::dxy, dz without a need for a pseudo-track.
Can this be reduced here as well?

Reply:
as we are already using the pseudotrack in the requirements above, we would prefer to keep the cutflow consistent with the current definition